[Review] TodoList Review 페이지 입니다.

seyoon-kim opened the pull request:
**데모페이지**
[데모페이지 링크](http://10.78.65.35:8000/example/main.html)

**요구사항**
[요구사항 링크](https://github.nhnent.com/fe/training/wiki/TodoList-%EA%B5%AC%ED%98%84)

**소개**
Todolist를 개발하고 리뷰한다

**요구사항**

**기본 조건**

- 학습한 모든 개발 도구를 사용한다.(eslint, jasmine, karma, bower, webpack, gulp)
- 자바스크립트 라이브러리는 codesnippet만을 이용한다.

**구현 스펙**

- UI는 Todo 입력부와 Todo 목록 출력부, 정보 출력부로 나뉜다.
- Todo는 [완료 전], [완료] 두 가지 상태를 가진다. (default : 완료 전)

**TODO 입력부**

- Todo를 입력 받을 수 있는 input 있다.
- Input에 텍스트로 Todo를 입력할 수 있으며, enter키로 Todo를 등록할 수 있다.
- 등록된 Todo는 Todo 목록 상단에 추가되며, 등록과 동시에 input의 내용은 초기화된다.

**TODO 목록 출력부**

- 등록된 Todo 목록이 출력된다.
- Todo 는 등록순으로 소팅되어 최근 목록이 상단에 위치한다.
- Todo는 아래와 같이 구성된다.
- 완료 여부를 나타내는 checkbox
- 내용을 나타내는 textnode
- checkbox를 클릭하여 완료 처리할 수 있으며, 토클 방식으로 상태를 변경할 수 있다.
- 체크 : [완료] 상태
- 체크해제 : [완료 전] default 상태
- 완료된 Todo는 Todo 목록의 하단으로 이동하며, 이미 완료된 Todo를 다시 [완료 전] 상태로 변경하면 Todo 목록 상단으로 다시 이동한다.
- 완료된 Todo는 취소선과 폰트 컬러로 시각적으로 다르게 한다.

**하단 정보 출력부**

- 현재 남아있는 완료 전 Todo의 갯수를 출력한다.
- [전체], [완료 전], [완료됨] 의 탭으로 Todo목록을 필터해 볼 수 있는 기능을 제공한다.
- 완료항목 삭제 기능을 제공한다.

seyoon-kim committed:
feat, loadTodoData, renderView 함수구현

-loadTodoData, 데이터를 받아서 isCheck 값 별로 배열로 저장한다
- renderView 저장된 데이터값을 확인하여 todoList화면을 렌더링한다.

seyoon-kim committed:
feat, todo의 checkbox를 클릭하는 경우 이벤트 구현

todo의 checkbox를 클릭 하는 경우 해당 todo의 isChecked값이 true이면 false로 바뀐뒤 해당 todo가 삭제되고 incompleteList 엘리멘트의 첫번째 자식으로 들어가게 된다.

seyoon-kim committed:
feat, infoList에서 현재 미완료된 todo가 몇개인지 보여주기 구현

하단에 정보를 보여주는 infoList를 생성하고 그 중에서 현재 미완료된 todo의 갯수가 몇개인지 데이터가 바뀔때마다 바뀌도록 구현

seyoon-kim committed:
feat, infoList 내에서 현재 완료된 Todo의 갯수가 몇개인지 알려주는 버튼구현

- 현재 완료된 Todo의 갯수가 몇개인지 알려주는 버튼 구현
- 해당 버튼 클릭시 현재 완료된 Todo, completeList에서 모두 삭제

seyoon-kim committed:
feat, todoList filter기능을 갖는 버튼을 구현

- btnAllList, complteList, incompleteList 둘다 보여줌
- btnAtiveList, complteList 안보여주고 incompleteList 보여줌
- btnComplteList, complteList 보여주고 incomplteList 안보여줌

seyoon-kim committed:
refactor, 소스 리팩토링

1. 변수명 좀더 구체적으로 변경
2. 중복되는 코드는 함수를 만들어 코드축소
3. 함수 설명 주석추가
4. 초기화 안된 변수는 한줄로 변경

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+        key = e.keyCode;
+        targetId = target.getAttribute('id');
+
+        if (target.value.length === 0) {
```
`=== 0` 비교는 그냥 생략하는 게 팀 컨벤션입니다. 
`if (target.value.length)` 라고만 해도 됩니다. 
여기서는 length 도 필요 없이 `if (target.value)` 로만 체크해도 될 것 같네요.

seyoon-kim replied:
넵 '=== 0' 비교는 생략 하도록 하겠습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+        todoObjects = todoData;
+
+        completeTodoObjects = snippet.filter(todoObjects, function(value) {
+            return (value.isChecked === true);
```
`boolean`은 불필요하게 다시 비교할 필요 없어요.

seyoon-kim replied:
넵 비교구문 삭제하여 수정하였습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+        _sortRegDateOfTodoObject(completeTodoObjects);
+
+        incompleteTodoObjects = snippet.filter(todoObjects, function(value) {
+            return (value.isChecked === false);
```
여기도 `return !value.isChecked` 가 더 간결하게 좋겠네요.

seyoon-kim replied:
넵 말씀해주신 부분이 더간결하여 그렇게 수정하였습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+
+    var _renderInfoList = function() {
+        var eleLeftItemsNum = document.getElementById('leftItemsNum');
+        var eleCompleteItemsNum = document.getElementById('completeItemsNum');
```
꼭 그런건 아닌데. HTML과 Javascript와의 연결고리들은 각각의 함수 내부에 숨어있기 보다 한곳에 모여있는 게 관리하기 편합니다.
예를 들면 `leftItemsNum`, `completeItemNum` 같은 ID나 className 값들은 모두 상수로 빼서 한곳에서 관리하면 좋을 것 같아요.

seyoon-kim replied:
상수로 생각되어지는 것들은 한곳에 모아 함께 관리할 수 있도록 하겠습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+    var _toggleTodo = function(idTodo, isCheckedTarget) {
+        _forEach(todoObjects, function(index, value) {
+            if (todoObjects[index].id === idTodo) {
+                todoObjects[index].isChecked = (!!isCheckedTarget);
```
`forEach`를 사용하는데 `todoObjects[index]` 처럼 다시 index로 접근하는 이유를 모르겠네요. 
그냥 `value`를 사용하시면 됩니다.

seyoon-kim replied:
넵 value값을 사용하여 수정하였습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+        var target, eleTodo, idTodo;
+        e = e || window.event;
+        target = e.target || e.srcElement;
+        eleTodo = target.parentElement;
```
이런 크로스브라우징 처리는 유틸 함수로 빼서 관리해 주세요.

seyoon-kim replied:
넵 크로스브라우징 처리는 따로 관리하도록하겠습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+        var eleCompleteList = document.getElementById('completeList');
+        var eleIncompleteList = document.getElementById('incompleteList');
+        Domclass.removeClass(eleCompleteList, 'hide');
+        Domclass.removeClass(eleIncompleteList, 'hide');
```
Domclass는 전역에 있는 모듈인가요? 이 모듈도 CommonJS 형태로 사용하면 좋을 것 같습니다.

seyoon-kim replied:
현재 util.js안에 있는 Domclass, Domutil, Eventutil을 util이라는 폴더를 생성하고 각각의 파일을 이름에 맞게 생성하여 나누어 두었습니다.

dongwoo-kim commented on src/util.js:
```diff
@@ -1,6 +1,5 @@
 
 module.exports = Domutil = (function() {
```
이런 식으로 모듈을 전역으로 빼는 이유가 특별히 있는 건가요?

seyoon-kim replied:
현재 util.js안에 있는 Domclass, Domutil, Eventutil을 util이라는 폴더를 생성하고 각각의 파일을 이름에 맞게 생성하여 나누어 두었습니다.

dongwoo-kim commented on test/spec.js:
```diff
@@ -1,5 +1,171 @@
+    ];
+
+    beforeEach(function() {
+        document.body.innerHTML = '<div class="todoListWrap" id="todoListWrap"><div class="inputArea"><input type="text" class="inputTxt"/></div><div class="listArea"><ul class="incompleteList" id="incompleteList"></ul><ul class="completeList" id="completeList"></ul></div></div><div class="infoArea"><p class="leftItems"> <span id="leftItemsNum"></span> items left </p><div class="filterArea"><butt ...
```
이렇게 한줄로 써 놓으면.. 코드를 알아보기가 너무 어렵습니다.
적절하게 줄바꿈을 해서 잘 알아볼 수 있도록 변경해주세요.

dongwoo-kim replied:
그리고 이런 코드들은 fixture를 사용해서 html 페이지 형태로 관리하는 게 더 좋을 것 같아요.
https://github.com/velesin/jasmine-jquery 참고해 주세요

seyoon-kim replied:
fixture를 사용하여 test/fixtures 폴더를 따로 만들어 관련된 fixture들을 관리하고 사용하도록 변경하였습니다.

dongwoo-kim commented on test/spec.js:
```diff
@@ -1,5 +1,171 @@
+    it('completeList에 해당 엘리멘트가 추가되었는지 확인', function() {
+        var eleCompleteList = Domutil.querySelector('.completeList');
+        var eleLi = '<li class="todo" data-id="todo1"><input type="checkbox" class="todoChk" checked=""><p class="todoTitle">제목02</p></li>';
+        expect(eleCompleteList.innerHTML).toEqual(eleLi);
```
이런식으로 `innerHTML` 값을 직접 비교하게 되면 브라우저에 따라 실패하는 경우가 있을 것 같네요..

seyoon-kim replied:
직접 비교하지 않고 생성된 엘리멘트의 아이디값이나 클래스값의 존재 여부로 해당 엘리멘트가 있는지 비교하였습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+    };
+
+    /**
+     * 인자로 받은 데이터 정보를 todoObjects에 담고, isChecked = false인것은 incompleteTodoObjects, isChecked = true인것은 completeTodoObjects 배열에 담은 뒤 오름차순 정렬
```
~~ESLint의 max-len 설정이 100 정도로 되어 있을텐데.. 한 줄이 긴 경우게 많이 보이네요.~~

dongwoo-kim replied:
max-len 설정은 120인데. comment 나 string 은 무시하도록 되어 있네요.
ESLint 룰이 아니더라도, 주석도 적절하게 보기좋게 줄바꿈 해 주세요.

seyoon-kim replied:
긴 주석은 줄바꿈으로 바꾸도록 하겠습니다.

dongwoo-kim commented on npm-debug.log:
```diff
@@ -1,45 +0,0 @@
-23 error Or if that isn't available, you can get their info via:
-23 error     npm owner ls selector
-23 error There is likely additional logging output above.
-24 verbose exit [ 1, true ]
```
요런건 commit 안되게 .gitignore 처리 해주세요.

seyoon-kim replied:
넵 .gitignore 처리 했습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+            if (todoObjects[index].id === idTodo) {
+                todoObjects[index].isChecked = (!!isCheckedTarget);
+                _loadTodoData(todoObjects);
+                _renderView();
```
순회를 할 때마다 매번 렌더링을 할 필요는 없을 것 같네요. `_forEach` 바깥에서 렌더링 하면 되지 않을까요.

seyoon-kim replied:
렌더링 하는 부분은 forEach가 종료된 이후에 하도록 변경하였습니다.

dongwoo-kim commented on src/todoList.js:
```diff
@@ -1,2 +1,230 @@
+            return;
+        }
+
+        if (key === 13 && targetId === 'todoInputTxt') {
```
이런 키-코드들은 상수로 분리하는 게 좋아요.

seyoon-kim replied:
상수로 분리하였습니다.

dongwoo-kim reviewed (changes requested):
완료. 고생했어요. 카르마 웹드라이브 설정이랑 Eslint 수정까지 같이 반영해주세요 :)

seyoon-kim committed:
refactor, 리뷰 1차 정리

seyoon-kim committed:
refactor, 선임님 리뷰 반영 및 코드정리

1. karma webdriver속성을 이용하여 ie,ff, chrome, safari등 다양한 환경에서 테스트 할 수있도록 설정
2. util파일 Domutil, Domclass, Eventutil 세개의 파일로 분리
3. 테스트 파일 엘리멘트 작성 jasmine-jquery의 fixtures를 이용하여 보기 편하게 분리
4. 코드 소스 정리 등

seyoon-kim committed:
Refactor, module수정, 상수관리

1. module파일 내에 즉시실행함수로 감싸져있는 부분을 꺼내서 사용, Common.js로 module화 되어있는데 또 다시 즉시실행함수로 스코프를 생성할 필요가없다
2. 상수 정리, KEY_ENTER와 같이 고정되어있는 값이 아니면 상수라 할 수없다. todoListWrapElement의 경우 해당 아이디값에 따라 반환되는 객체가 바뀔수도 있기 때문에 상수라 보기 어렵다. 오히려 todoListWrap과 같은 ID값을 상수로 따로 관리하는게 맞다.

seyoon-kim commented:
선임님 리뷰해주셔서 감사합니다.
해당 PR은 리뷰가 종료되어 close 합니다.
